### PR TITLE
Docket Switch | Ability to Add Admin Action Tasks

### PR DESCRIPTION
**Pull request by jcq**
Connects #15100

### Description
This adds the ability to add admin action tasks during step 2 of docket switch (granted) checkout.

### Acceptance Criteria
- [ ] Code compiles correctly
- [ ] User can select one or more admin actions to add during docket switch checkout flow

### CODE REVIEW

File: client/app/queue/docketSwitch/denial/DocketSwitchDenialForm.jsx [hunk 1]
```diff
@@ -2,7 +2,7 @@ import React from 'react';
 import PropTypes from 'prop-types';
 
 import { useForm } from 'react-hook-form';
-import { yupResolver } from '@hookform/resolvers';
+import { yupResolver } from '@hookform/resolvers/yup';
```
Comment (jcq) [inline client/app/queue/docketSwitch/denial/DocketSwitchDenialForm.jsx:5]: Unfortunately, the imports for the yup resolver changed with the updated version, so there are a number of otherwise-unrelated files being updated here

File: client/app/queue/docketSwitch/docketSwitchSlice.js [hunk 2]
```diff
@@ -10,38 +9,36 @@ const initialState = {
...
+    cancel: () => ({ ...initialState }),
+    stepForward: (state) => ({ ...state, step: state.step + 1 }),
     stepBack: (state) => ({ ...state, step: state.step ? state.step - 1 : 0 }),
     updateDocketSwitch: (state, action) => {
-
       const { formData: updates } = action.payload;
-      updates.receiptDate = updates.receiptDate?.toISOString() ?? undefined; 
-      
+
+      if (updates.receiptDate) {
```
Comment (jcq) [inline client/app/queue/docketSwitch/docketSwitchSlice.js:29]: ensure we don't overwrite already selected receiptDate 😅

File: client/app/queue/docketSwitch/grant/DocketSwitchEditTasksContainer.jsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import React, { useMemo } from 'react';
```
Comment (jcq) [inline client/app/queue/docketSwitch/grant/DocketSwitchEditTasksContainer.jsx:1]: Renamed some files to make it clear that we are doing more than adding tasks (hence the "edit" nomenclature)

File: client/app/queue/docketSwitch/grant/DocketSwitchReviewRequestForm.jsx [hunk 4]
```diff
@@ -33,38 +33,56 @@ const schema = yup.object().shape({
...
+    () =>
+      Object.values(DISPOSITIONS).filter(
+        (disposition) => disposition.value !== 'denied'
+      ),
+    []
+  );
 
   const watchDisposition = watch('disposition');
 
+  // Ensure that we trigger revalidation whenever disposition changes
```
Comment (jcq) [inline client/app/queue/docketSwitch/grant/DocketSwitchReviewRequestForm.jsx:81]: Other than linting, this is the only change to this file. Not sure how tests were passing before, but this ensures that validation gets triggered any time the disposition changes. This allows it to properly disable button whenever selection is switched to "partially_granted" and no issues are selected.

File: client/app/queue/docketSwitch/judgeRuling/DocketSwitchRulingContainer.jsx [hunk 5]
```diff
@@ -3,7 +3,7 @@ import { useDispatch, useSelector } from 'react-redux';
 import { useHistory, useParams, useRouteMatch } from 'react-router';
 import { taskById, appealWithDetailSelector } from 'app/queue/selectors';
 import { taskActionData } from 'app/queue/utils';
-import DISPOSITIONS from 'constants/DOCKET_SWITCH';
+import DISPOSITIONS from 'constants/DOCKET_SWITCH_DISPOSITIONS';
```
Comment (jcq) [inline client/app/queue/docketSwitch/judgeRuling/DocketSwitchRulingContainer.jsx:6]: Renamed this because we added another file relating to Docket Switch

File: client/app/queue/docketSwitch/grant/DocketSwitchReviewRequestForm.jsx (hunk 4 above)
Comment (leikkisa) [reply client/app/queue/docketSwitch/grant/DocketSwitchReviewRequestForm.jsx:81]: This might explain a flakey Jest test failure I couldn't explain on the claimant work, which was passing locally but failing in Circle (I ended up skipping the test)

File: client/app/queue/docketSwitch/grant/DocketSwitchEditTasksForm.jsx [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+      ...item,
+      disabled: true,
+    }));
+  }, [docketTo]);
+
+  // Preselect all "mandatory" task options
+  useEffect(() => {
+    const newValues = {};
+
+    mandatoryTaskOptions.forEach((item) => (newValues[item.id] = true));
```
Comment (leikkisa) [inline client/app/queue/docketSwitch/grant/DocketSwitchEditTasksForm.jsx:96]: I added a comment below that calling the type/label and ID is a little confusing to me. I know it might interact with previous work so not sure how hard it is, but I would say it's low priority / non-blocking.

File: client/constants/DOCKET_SWITCH_TASKS_BY_DOCKET_TYPE.json [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    {
+      "id": "EvidenceSubmissionWindowTask",
+      "label": "Evidence Submission Window Task"
+    }
+  ],
+  "hearing": [
+    { "id": "DistributionTask", "label": "Distribution Task" },
+    { "id": "HearingTask", "label": "Hearing Task" }
+  ]
+}
```
Comment (leikkisa) [inline client/constants/DOCKET_SWITCH_TASKS_BY_DOCKET_TYPE.json:16]: Love this approach. I think "id" is a little counterintuitive (it's more like a task type or label). I would say this is non-blocking, but if it's not hard to update the naming I think it would make more sense, but again low priority.

File: spec/feature/queue/docket_switch_spec.rb [hunk 8]
```diff
@@ -347,6 +351,38 @@
...
+      end
+
+      # Add new Admin Action
+      click_button("+ Add task")
+      expect(page).to have_button("Continue", disabled: true)
+
+      # Ensure all admin actions are available and select "AOJ"
+      click_dropdown(text: "AOJ") do
+        visible_options = page.find_all(".cf-select__option")
+        expect(visible_options.length).to eq Constants::CO_LOCATED_ADMIN_ACTIONS.length
```
Comment (leikkisa) [inline spec/feature/queue/docket_switch_spec.rb:374]: Nice

File: client/app/queue/docketSwitch/grant/DocketSwitchAddAdminTaskForm.jsx [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+  return (
+    <>
+      <div className={fieldStyles}>
+        <Controller
+          name={`${baseName}.type`}
+          control={control}
+          defaultValue={item.type}
+          render={({ onChange, ...rest }) => (
+            <SearchableDropdown
```
Comment (leikkisa) [inline client/app/queue/docketSwitch/grant/DocketSwitchAddAdminTaskForm.jsx:44]: It's very cool to see options for how this can be used (after running into this with the appellant work)

File: client/app/queue/docketSwitch/grant/DocketSwitchAddAdminTaskForm.jsx [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+};
+
+DocketSwitchAddAdminTaskForm.propTypes = {
+  baseName: PropTypes.string.isRequired,
+  item: PropTypes.shape({
+    type: PropTypes.shape({ value: PropTypes.string, label: PropTypes.string }),
+    instructions: PropTypes.string,
+  }),
+  onRemove: PropTypes.func.isRequired,
+};
```
Comment (leikkisa) [inline client/app/queue/docketSwitch/grant/DocketSwitchAddAdminTaskForm.jsx:81]: Nicely done

File: client/app/queue/docketSwitch/grant/DocketSwitchAddAdminTaskForm.stories.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    onRemove: { action: 'onRemove' },
+  },
+};
+const Template = (args) => <DocketSwitchAddAdminTaskForm {...args} />;
+
+export const Basic = Template.bind({});
+Basic.parameters = {
+  docs: {
+    storyDescription:
+      'Used by attorney in Clerk of the Board office to complete a grant of a docket switch checkout flow ',
```
Comment (leikkisa) [inline client/app/queue/docketSwitch/grant/DocketSwitchAddAdminTaskForm.stories.js:30]: This description looks the same for the review request form, edit task form and add admin task form. Not sure if it would make sense to be more specific. Non-blocking.

File: client/app/queue/docketSwitch/grant/DocketSwitchEditTasksContainer.jsx (hunk 3 above)
Comment (leikkisa) [reply client/app/queue/docketSwitch/grant/DocketSwitchEditTasksContainer.jsx:1]: Love the renaming, yeah before it was less intuitive but "edit" covers the full functionality.

File: client/app/queue/docketSwitch/grant/DocketSwitchEditTasksContainer.jsx [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    return tasks.filter((task) => task.canMoveOnDocketSwitch);
+  }, [tasks]);
+
+  const handleSubmit = (formData) => {
+    // Add stuff to redux store
+    dispatch(updateDocketSwitch({ formData }));
+
+    // Move to next step
+    dispatch(stepForward());
+    push(`/queue/appeals/${appealId}/tasks/${taskId}/docket_switch/checkout/grant/confirm`);
```
Comment (leikkisa) [inline client/app/queue/docketSwitch/grant/DocketSwitchEditTasksContainer.jsx:49]: Very exciting

File: client/app/queue/docketSwitch/grant/DocketSwitchEditTasksForm.jsx [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+
+    taskListing.forEach((item) => (newValues[item.taskId] = true));
+    setTasks(newValues);
+
+    setValue('taskIds', taskListing.map((task) => task.taskId));
+  }, [taskOptions]);
+
+  // Used for display of mandatory tasks
+  const mandatoryTaskOptions = useMemo(() => {
+    return tasksByDocketType[docketTo].map((item) => ({
```
Comment (leikkisa) [inline client/app/queue/docketSwitch/grant/DocketSwitchEditTasksForm.jsx:86]: Nice

**Review by leikkisa — APPROVED** [review]
Hey JC, I'm very sorry for the delay. This PR looks very good, you did a lot of work. Thank you (as always) for such doing such high quality work, continuously improving code, and being flexible with the scope changes that I think are a big improvement. 💯 

I left some minor non-blocking comments, and they're no big deal but just things I noticed.

**Comment by leikkisa** [comment]
Also I'm not sure if we still need to add screenshots somewhere else, but I think it would be good to add them to the PR. I tested multiple scenarios locally (looks so good!)

File: client/constants/DOCKET_SWITCH_TASKS_BY_DOCKET_TYPE.json (hunk 7 above)
Comment (jcq) [reply client/constants/DOCKET_SWITCH_TASKS_BY_DOCKET_TYPE.json:16]: Yeah, I did this just because it meant we could use it as-is for `<SearchableDropdown>`, but we could always just rename them once after reading as well

File: client/app/queue/docketSwitch/grant/DocketSwitchAddAdminTaskForm.stories.js (hunk 11 above)
Comment (jcq) [reply client/app/queue/docketSwitch/grant/DocketSwitchAddAdminTaskForm.stories.js:30]: I'll maybe tweak these in my final Review/Confirm PR

File: client/app/queue/docketSwitch/grant/DocketSwitchEditTasksForm.jsx (hunk 6 above)
Comment (jcq) [reply client/app/queue/docketSwitch/grant/DocketSwitchEditTasksForm.jsx:97]: Yup yup — went ahead and renamed the object keys
